src/Validation/Validator.php

```diff
@@ -430,7 +430,7 @@ public function count(): int
      * @param array|\Cake\Validation\ValidationRule $rule the rule to add
      * @return $this
      */
-    public function add(string $field, $name, $rule = [])
+    public function add(string $field, $name, $rule = []): self
```

Comment:
We must never add self etc for chainable. You shoot yourself in the foot here.
There is also no need to protect anyone, we all know what return $this in docblock means, the IDE does too.
Code sniffer takes care about the rest.